src/shogun/mathematics/linalg/ratapprox/logdet/LogDetEstimator.cpp

```diff
@@ -148,72 +125,81 @@ SGVector<float64_t> CLogDetEstimator::sample(index_t num_estimates)
 		m_operator_log->get_operator()->get_dimension(),
 		m_trace_sampler->get_dimension());
 
-	// for storing the aggregators that submit_jobs return
-	CDynamicObjectArray* aggregators=new CDynamicObjectArray();
+	// for storing the result
+	CDynamicArray<float64_t> aggregators;
 	index_t num_trace_samples=m_trace_sampler->get_num_samples();
-
-	for (index_t i=0; i<num_estimates; ++i)
+	// for omp
+	int32_t num_vectors = num_estimates;
+	auto pb = progress(range(num_vectors));
+	int32_t num_threads;
+	int64_t step;
+#pragma omp parallel shared(num_threads, step)
 	{
-		for (index_t j=0; j<num_trace_samples; ++j)
-		{
-			SG_INFO("Computing log-determinant trace sample %d/%d\n", j,
-					num_trace_samples);
 
-			SG_DEBUG("Creating job for estimate %d, trace sample %d/%d\n", i, j,
+#ifdef HAVE_OPENMP // HAVE OPENMP
+#pragma omp single
+		{
+			num_threads = omp_get_num_threads();
+			step = num_vectors / num_threads;
+			num_threads--;
+		}
+		int32_t thread_num = omp_get_thread_num();
+#else
+		num_threads = 0;
+		step = num_vectors;
+		int32_t thread_num = 0;
+#endif
+		int32_t start = thread_num * step;
+		int32_t end =
+			(thread_num == num_threads) ? num_vectors : (thread_num + 1) * step;
+
+#ifdef WIN32 // HAVE WIN32
+		for (int32_t vec = start; vec < end; vec++)
+#else
```

Comment:
what is the purpose of this contraption?

Reply:
In case if openmp is not supported this will work like a serial implementation i found a similar thing at shogun ```shogun/features/DotFeatures.cpp```

Reply:
I think a better would be to look at something like this:
https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/regression/LeastAngleRegression.cpp#L251

Reply:
Interesting.... this is easier i will use that for sure.